File: client/src/utils/protectedText/patternRenderer.ts
/**
 * PATTERN rendering strategy — Protected Text v2.
 *
 * Shows only a coarse, locally-classified intent hint as an abstract glyph sequence. The actual
 * plaintext is NEVER encoded — only the classified `IntentCategory` plus decorative, seed-derived
 * filler glyphs pulled from a small fixed alphabet.
 *
 * Fixed grammar (spec §14): PREFIX + VISUAL_TOKEN + SEPARATOR + INTENT_SYMBOL + SEPARATOR + VISUAL_TOKEN
 * Deterministic: same (plaintext, rendererVersion) always renders the same glyph sequence, so the
 * message doesn't flicker between renders while displayed.
 */

import { deriveRenderSeed } from './sha256';
import { classifyIntent, type IntentCategory } from './intentClassifier';

const PREFIXES = ['⟐', '◈', '❖', '✦'];
const SEPARATOR = '·';

// Small fixed alphabet used to build decorative, seed-derived visual tokens — deliberately
// styled to resemble the ILLUSION glyph set so PATTERN output reads as part of the same
// visual language, without encoding anything about the message content.
const TOKEN_ALPHABET = 'qwΜηєℓ7Xv4Ζσ9τяkPb';

const INTENT_SYMBOLS: Record<IntentCategory, string> = {
  URGENT: '‼',
  QUESTION: '?',
  TIME: '○',
  LOCATION: '⟐',
  REQUEST: '→',
  NEGATION: '-',
  AFFIRMATION: '+',
  GREETING: '~',
  FAREWELL: '»',
  ACKNOWLEDGEMENT: '✓',
  GENERAL: '•',
};

function buildVisualToken(seed: Uint8Array, offset: number, length = 3): string {
  let token = '';
  for (let i = 0; i < length; i++) {
    const byte = seed[(offset + i) % seed.length];
    token += TOKEN_ALPHABET[byte % TOKEN_ALPHABET.length];
  }
  return token;
}

/**
 * Renders a message as a PATTERN-mode intent hint. Deterministic and content-blind beyond the
 * coarse classified category.
 */
export function renderPattern(content: string, rendererVersion: number): string {
  if (!content) return '';

  const seed = deriveRenderSeed(content, rendererVersion, 'PATTERN');
  const intent = classifyIntent(content);

  const prefix = PREFIXES[seed[0] % PREFIXES.length];
  const symbol = INTENT_SYMBOLS[intent];
  const tokenA = buildVisualToken(seed, 1);
  const tokenB = buildVisualToken(seed, 4);

  return `${prefix} ${tokenA} ${SEPARATOR} ${symbol} ${SEPARATOR} ${tokenB}`;
}

export { INTENT_SYMBOLS };

File: android/app/src/main/java/com/enctxt/core/privacy/PatternRenderer.kt
package com.enctxt.core.privacy

/**
 * PATTERN rendering strategy — Protected Text v2.
 *
 * Shows only a coarse, locally-classified intent hint as an abstract glyph sequence. The actual
 * plaintext is NEVER encoded — only the classified [IntentCategory] plus decorative, seed-derived
 * filler glyphs pulled from a small fixed alphabet.
 *
 * Fixed grammar (spec §14): PREFIX + VISUAL_TOKEN + SEPARATOR + INTENT_SYMBOL + SEPARATOR + VISUAL_TOKEN
 * Must match the Web implementation (client/src/utils/protectedText/patternRenderer.ts) exactly.
 */
object PatternRenderer {

    private val PREFIXES = listOf("⟐", "◈", "❖", "✦")
    private const val SEPARATOR = "·"
    private const val TOKEN_ALPHABET = "qwΜηєℓ7Xv4Ζσ9τяkPb"

    val INTENT_SYMBOLS: Map<IntentCategory, String> = mapOf(
        IntentCategory.URGENT to "‼",
        IntentCategory.QUESTION to "?",
        IntentCategory.TIME to "○",
        IntentCategory.LOCATION to "⟐",
        IntentCategory.REQUEST to "→",
        IntentCategory.NEGATION to "-",
        IntentCategory.AFFIRMATION to "+",
        IntentCategory.GREETING to "~",
        IntentCategory.FAREWELL to "»",
        IntentCategory.ACKNOWLEDGEMENT to "✓",
        IntentCategory.GENERAL to "•"
    )

    private fun buildVisualToken(seed: ByteArray, offset: Int, length: Int = 3): String {
        val sb = StringBuilder(length)
        for (i in 0 until length) {
            val byte = seed[(offset + i) % seed.size].toInt() and 0xFF
            sb.append(TOKEN_ALPHABET[byte % TOKEN_ALPHABET.length])
        }
        return sb.toString()
    }

    /**
     * Renders a message as a PATTERN-mode intent hint. Deterministic and content-blind beyond
     * the coarse classified category.
     */
    fun protect(content: String, rendererVersion: Int): String {
        if (content.isEmpty()) return ""

        val seed = RenderSeed.derive(content, rendererVersion, "PATTERN")
        val intent = IntentClassifier.classify(content)

        val prefix = PREFIXES[(seed[0].toInt() and 0xFF) % PREFIXES.size]
        val symbol = INTENT_SYMBOLS.getValue(intent)
        val tokenA = buildVisualToken(seed, 1)
        val tokenB = buildVisualToken(seed, 4)

        return "$prefix $tokenA $SEPARATOR $symbol $SEPARATOR $tokenB"
    }
}
